torchdata/nodes/map.py

```diff
@@ -0,0 +1,132 @@
+# pyre-unsafe
+# import multiprocessing as mp
+import queue
+import threading
+from typing import Callable, Iterator, List, Literal, Optional, TypeVar, Union
+
+import torch.multiprocessing as mp
+
+from torch._utils import ExceptionWrapper
+
+from torchdata.nodes import BaseNode, T
+
+from ._apply_udf import _apply_udf
+
+from ._populate_queue import _populate_queue
+
+
+X = TypeVar("X")
+
+
+class Mapper(BaseNode[T]):
+    def __init__(
+        self,
+        source: BaseNode[X],
+        map_fn: Callable[[X], T],
+    ):
+        self.source = source
+        self.map_fn = map_fn
+
+    def iterator(self) -> Iterator[T]:
+        for item in self.source:
+            yield self.map_fn(item)
+
+
+class ParallelMapper(BaseNode[T]):
+    def __init__(
+        self,
+        source: BaseNode[X],
+        map_fn: Callable[[X], T],
+        num_workers: int,
+        in_order: bool = True,
+        method: Literal["thread", "process"] = "thread",
+    ):
+        self.source = source
+        self.udf = map_fn
+        self.num_workers = num_workers
+        self.in_order = in_order
+
+        self.in_q: Union[queue.Queue, mp.Queue] = queue.Queue() if method == "thread" else mp.Queue()
+        self.out_q: Union[queue.Queue, mp.Queue] = queue.Queue() if method == "thread" else mp.Queue()
+        self.sem = threading.BoundedSemaphore(value=2 * num_workers)
+
+        self._started = False
+        self._stop = threading.Event()
+        self._mp_stop = mp.Event()
+        self._read_thread = threading.Thread(
+            target=_populate_queue, args=(self.source, self.in_q, self._stop, self.sem)
+        )
+        self._method = method
+
+        self._map_threads: List[Union[threading.Thread, mp.Process]] = []
+
+    def iterator(self) -> Iterator[T]:
+        if not self._started:
+            for worker_id in range(self.num_workers):
+                args = (
+                    worker_id,
+                    self.in_q,
+                    self.out_q,
+                    self.in_order,
+                    self.udf,
+                    self._stop if self._method == "thread" else self._mp_stop,
+                )
+                self._map_threads.append(
+                    threading.Thread(target=_apply_udf, args=args)
+                    if self._method == "thread"
+                    else mp.Process(target=_apply_udf, args=args)
+                )
```

Comment:
For all of the other thread object creation, we do it in the constructor. Is there a particular reason that we defer the thread object creation for the map threads until the time we start all the threads?

Reply:
No good reason other than it was the way the code evolved during experimentation. Let's make sure there's uniformity before landing, and also in variable naming conventions

Reply:
Just a follow-up: we'll consolidate as part of iterator creation, this is how it's handled in current torch.utils.data.DataLoader and part of this is so that IterableDataset's have a known time (iterator creation" when they will be instantiated inside a process, so they can shard themselves. We'll keep this to make sure we don't run into BC issues later